qiskit/circuit/library/generalized_gates/permutation.py

```diff
@@ -77,17 +83,59 @@ def __init__(
             pattern = np.arange(num_qubits)
             rng.shuffle(pattern)
 
-        name = "permutation_" + np.array_str(pattern).replace(" ", ",")
+        # This is needed to support qasm()
+        self._qasm_name = "permutation__" + "_".join([str(n) for n in pattern]) + "_"
+        self._qasm_definition = None
 
-        circuit = QuantumCircuit(num_qubits, name=name)
+        super().__init__(name="permutation", num_qubits=num_qubits, params=[pattern])
 
-        super().__init__(num_qubits, name=name)
+    def __array__(self, dtype=None):
+        """Return a numpy.array for the Permutation gate."""
+        nq = len(self.pattern)
+        mat = np.zeros((2**nq, 2**nq), dtype=dtype)
+
+        for r in range(2**nq):
+            # convert row to bitstring, reverse, apply permutation pattern, reverse again,
+            # and convert to row
+            bit = bin(r)[2:].zfill(nq)[::-1]
+            permuted_bit = "".join([bit[j] for j in self.pattern])
+            pr = int(permuted_bit[::-1], 2)
+            mat[pr, r] = 1
+
+        return mat
+
+    def validate_parameter(self, parameter):
```

Comment:
why is this needed? what is the `parameter`?

Reply:
This is unfortunately needed. `Permutation` inherit from `Gate` and `parameter` is whatever is in `params`, see this line in `Permutation.__init__`: `super().__init__(name="permutation", num_qubits=num_qubits, params=[pattern])`, i.e. for permutations the only parameter is the permutation pattern. The `Gate` class calls `validate_parameter` which does not like arrays, and the common strategy is to override this method in the child class (e.g., we do the same in `LinearFunction`).